File: src/bank/camt.rs
// bukio-cli — agent-first double-entry bookkeeping for SMEs.
// Copyright (c) 2026 Erik van Kempen.
// SPDX-License-Identifier: Apache-2.0
//
// CAMT.053 XML parser (mirrors src/bank/camt.js).

use crate::bank::BankTx;
use crate::money::BukioError;

/// Parse CAMT.053 XML into a list of bank transactions.
/// Uses simple string scanning (no XML crate needed for the well-defined SDMX-like structure).
pub fn parse_camt053(xml: &str) -> Result<Vec<BankTx>, BukioError> {
    let mut transactions = Vec::new();

    // find all Ntry blocks
    let mut pos = 0;
    while let Some(ntry_start) = xml[pos..].find("<Ntry>") {
        let abs_start = pos + ntry_start;
        if let Some(ntry_end) = xml[abs_start..].find("</Ntry>") {
            let block = &xml[abs_start..abs_start + ntry_end + 7];
            if let Some(tx) = parse_ntry(block) {
                transactions.push(tx);
            }
            pos = abs_start + ntry_end + 7;
        } else {
            break;
        }
    }
    Ok(transactions)
}

fn parse_ntry(block: &str) -> Option<BankTx> {
    let date = extract_tag(block, "Dt")?;
    let amount_str = extract_tag(block, "Amt")?;
    let amount_cents = parse_amount_cents(&amount_str)?;
    let counterparty = extract_tag(block, "Nm").or_else(|| extract_tag(block, "Cdtr/Nm"));
    let description = extract_tag(block, "AddtlNtryInf")
        .or_else(|| extract_tag(block, "Ustrd"))
        .or_else(|| extract_tag(block, "Desc"));
    let iban_counter =
        extract_tag(block, "CdtrAcct/Id/IBAN").or_else(|| extract_tag(block, "DbtrAcct/Id/IBAN"));
    let bank_ref = extract_tag(block, "AcctSvcrRef");
    // DBIT = money out (negative), CRDT = money in (positive)
    let direction = extract_tag(block, "CdtDbtInd").unwrap_or_default();
    let signed_cents = if direction == "DBIT" {
        -amount_cents
    } else {
        amount_cents
    };

    Some(BankTx {
        date,
        amount_cents: signed_cents,
        counterparty,
        description,
        iban_counter,
        bank_ref,
    })
}

fn extract_tag(xml: &str, tag: &str) -> Option<String> {
    // find <tag (with or without attributes) followed by >
    let tag_prefix = format!("<{tag}");
    let start = xml.find(&tag_prefix)?;
    let after_tag = &xml[start + tag_prefix.len()..];
    let gt_pos = after_tag.find('>')?;
    let content_start = start + tag_prefix.len() + gt_pos + 1;
    let end_tag = format!("</{tag}>");
    let end = xml[content_start..].find(&end_tag)?;
    let value = &xml[content_start..content_start + end];
    let trimmed = value.trim().to_string();
    if trimmed.is_empty() {
        None
    } else {
        Some(trimmed)
    }
}

fn parse_amount_cents(s: &str) -> Option<i64> {
    // CAMT amounts: "1234.56" or "-1234.56"
    let parts: Vec<&str> = s.split('.').collect();
    if parts.len() == 2 {
        let whole: i64 = parts[0].parse().ok()?;
        let sign = if whole < 0 { -1 } else { 1 };
        let frac_str = parts[1];
        let frac: i64 = if frac_str.len() == 1 {
            frac_str.parse::<i64>().ok()? * 10
        } else if frac_str.len() >= 2 {
            frac_str[..2].parse().ok()?
        } else {
            0
        };
        Some(sign * (whole.abs() * 100 + frac))
    } else {
        let v: i64 = s.parse().ok()?;
        Some(v * 100)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_simple_camt() {
        let xml = r#"<?xml version="1.0" encoding="UTF-8"?>
<Document xmlns="urn:iso:std:iso:20022:tech:xsd:camt.053.001.08">
  <BkToCstmrStmt>
    <Stmt>
      <Ntry>
        <Amt Ccy="EUR">123.45</Amt>
        <CdtDbtInd>CRDT</CdtDbtInd>
        <BookgDt><Dt>2026-01-15</Dt></BookgDt>
        <NtryDtls><TxDtls>
          <Refs><AcctSvcrRef>TX001</AcctSvcrRef></Refs>
          <Cdtr><Nm>ACME Corp</Nm></Cdtr>
          <CdtrAcct><Id><IBAN>NL91ABNA0417164300</IBAN></Id></CdtrAcct>
          <AddtlNtryInf>Invoice 42</AddtlNtryInf>
        </TxDtls></NtryDtls>
      </Ntry>
      <Ntry>
        <Amt Ccy="EUR">50.00</Amt>
        <CdtDbtInd>DBIT</CdtDbtInd>
        <BookgDt><Dt>2026-01-16</Dt></BookgDt>
        <NtryDtls><TxDtls>
          <Refs><AcctSvcrRef>TX002</AcctSvcrRef></Refs>
          <Dbtr><Nm>Shop</Nm></Dbtr>
        </TxDtls></NtryDtls>
      </Ntry>
    </Stmt>
  </BkToCstmrStmt>
</Document>"#;
        let txs = parse_camt053(xml).unwrap();
        assert_eq!(txs.len(), 2);
        assert_eq!(txs[0].amount_cents, 12345);
        assert_eq!(txs[0].counterparty.as_deref(), Some("ACME Corp"));
        assert_eq!(txs[0].bank_ref.as_deref(), Some("TX001"));
        assert_eq!(txs[1].amount_cents, -5000);
    }
}

File: src/bank/camt.js
/**
 * bukio-cli — agent-first double-entry bookkeeping for SMEs across thirty-one jurisdictions.
 * Copyright (c) 2026 Erik van Kempen.
 * SPDX-License-Identifier: Apache-2.0
 */

// CAMT.053 (ISO 20022 bank statement) parser.
import { XMLParser } from 'fast-xml-parser';

export function bankError(code, message) {
  const e = new Error(message);
  e.code = code;
  return e;
}

/**
 * Parse a decimal string to integer cents (CAMT uses '.' decimals).
 * Integer-only (house rule: no float money math). ISO 20022 amounts allow
 * up to 5 fraction digits; a third+ digit rounds half-up, exactly matching
 * the old parseFloat(x)*100+Math.round semantics without the float path.
 */
function cents(value) {
  const s = String(value ?? '').trim();
  if (!s) return null;
  const m = s.match(/^(-?)(\d+)(?:\.(\d+))?$/);
  if (!m) return null; // not a plain decimal (rejects '1e5', '1.2.3', '€ 5')
  const negative = m[1] === '-';
  const whole = parseInt(m[2], 10);
  const frac = (m[3] ?? '').padEnd(2, '0').slice(0, 2);
  let c = whole * 100 + parseInt(frac || '0', 10);
  const rest = (m[3] ?? '').slice(2);
  if (rest && parseInt(rest[0], 10) >= 5) c += 1; // half-up on the 3rd decimal
  return negative && c !== 0 ? -c : c;
}

function text(node) {
  if (node == null) return null;
  if (typeof node === 'string') {
    // an EMPTY element (<Nm></Nm>) parses to '' — treat it as absent so the
    // counterparty/description fallback chains (??) fall through instead of
    // storing an empty string as the counterparty
    return node.trim() ? node : null;
  }
  return null;
}

function isoDate(node) {
  const s = text(node);
  if (!s) return null;
  return s.slice(0, 10); // strip timezone offset
}

/**
 * Parse CAMT.053 XML into transactions:
 * [{ date, amount_cents, counterparty, description, iban_counter }]
 * amount sign: CRDT = positive (money in), DBIT = negative (money out).
 */
export function parseCamt053(xmlText) {
  const parser = new XMLParser({
    ignoreAttributes: true,
    removeNSPrefix: true,
    parseTagValue: false,
    trimValues: true,
  });
  let doc;
  try {
    doc = parser.parse(xmlText);
  } catch (err) {
    throw bankError('INVALID_CAMT', `could not parse CAMT.053 XML: ${err.message}`);
  }

  const stmt = doc?.Document?.BkToCstmrStmt?.Stmt;
  if (!stmt) throw bankError('INVALID_CAMT', 'no BkToCstmrStmt/Stmt found in the XML');

  const stmts = Array.isArray(stmt) ? stmt : [stmt];
  const transactions = [];

  for (const s of stmts) {
    const iban = text(s?.Acct?.Id?.IBAN);
    if (!iban) throw bankError('INVALID_CAMT', 'statement is missing Acct/Id/IBAN');
    const entries = s?.Ntry;
    if (!entries) continue;
    const entryList = Array.isArray(entries) ? entries : [entries];

    for (const ntry of entryList) {
      const amount = cents(ntry?.Amt);
      if (amount == null) {
        // a statement whose Ntry has no parseable <Amt> is corrupt — silently
        // skipping it would import a partial statement and make the account
        // balance diverge from the bank's without any warning
        throw bankError('INVALID_CAMT', 'Ntry without a valid cbc:Amt — the statement is corrupt; fix it before importing');
      }
      const direction = String(ntry?.CdtDbtInd ?? '').toUpperCase();
      const date = isoDate(ntry?.BookgDt?.Dt ?? ntry?.ValDt?.Dt ?? ntry?.BookgDt?.Dbt);
      // The bank's own entry reference (AcctSvcrRef) is unique per entry and
      // goes into the dedup hash — two genuinely identical same-day payments
      // (same amount, same counterparty, same description) must both import.
      const bankRef = text(ntry?.AcctSvcrRef);
      const txDtls = ntry?.NtryDtls?.TxDtls;
      const txList = txDtls ? (Array.isArray(txDtls) ? txDtls : [txDtls]) : [{}];

      for (const tx of txList) {
        const rltd = tx?.RltdPties ?? {};
        // counterparty: the other party — creditor when money goes out, debtor when money comes in
        const counterparty = text(direction === 'DBIT' ? rltd?.Cdtr?.Nm : rltd?.Dbtr?.Nm)
          ?? text(rltd?.Dbtr?.Nm)
          ?? text(rltd?.Cdtr?.Nm);
        const ibanCounter = text(direction === 'DBIT' ? rltd?.Cdtr?.Acct?.IBAN : rltd?.Dbtr?.Acct?.IBAN);
        const ustrd = tx?.RmtInf?.Ustrd;
        const description = Array.isArray(ustrd) ? ustrd.filter(Boolean).join(' ') : (text(ustrd) ?? text(ntry?.AddtlNtryInf) ?? '');
        const signedAmount = direction === 'DBIT' ? -amount : amount;

        transactions.push({
          date,
          amount_cents: signedAmount,
          counterparty,
          description,
          iban_counter: ibanCounter ?? null,
          bank_ref: bankRef ?? null,
          iban,
        });
      }
    }
  }

  if (transactions.length === 0) {
    throw bankError('EMPTY_STATEMENT', 'no transactions found in the CAMT.053 statement');
  }
  return transactions;
}
